### PR TITLE
Add html form to gosh index page

### PR DESCRIPTION
**Pull request by sgfk**
* Rewrite index template
* Provide an html form to post files to gosh

image

### CODE REVIEW

File: internal/server.go [hunk 1]
```diff
@@ -11,50 +11,142 @@ import (
...
-
-
-## Abuse
-
-If, for whatever reason, you would like to have a file removed prematurely,
-please write an e-mail to <{{.EMail}}>.
-
-Please allow me a certain amount of time to react and work on your request.
-`
+const indexTpl = `<html>
```
Comment (oxzi) [inline internal/server.go:14]: Please add a doctype here.

File: internal/server.go [hunk 2]
```diff
@@ -11,50 +11,142 @@ import (
...
+$ curl -F 'file=@foo.png' -F 'time=1m' -F 'burn=1' {{.Proto}}://{{.Hostname}}/</pre>
+
+  <h3>### form</h3>
+  
+  <form action="{{.Proto}}://{{.Hostname}}/" method="POST" enctype="multipart/form-data">
+    <div id="grid">
+      <label for="file">Your file:</label>
+      <input type="file" name="file" />
+      <label for="burn">Burn after reading:</label>
+      <input type="text" name="burn" pattern="1?" title="Use '1' to burn, leave empty to retain your file" />
```
Comment (oxzi) [inline internal/server.go]: Wouldn't a checkbox here be more intuitive?

File: internal/server.go [hunk 3]
```diff
@@ -11,50 +11,142 @@ import (
...
+      <label for="time">Optionally, set a custom expiry date:</label>
+      <input
+        type="text"
+        name="time"
+        pattern="[-+]?([0-9]+(\.[0-9]*)?(ns|us|ms|s|m|h)+)+"
+        title="A duration string is a possibly signed sequence of decimal numbers,
```
Comment (oxzi) [inline internal/server.go]: The server does not parse a time.Duration, only a variant.

https://github.com/oxzi/gosh/blob/8185d2179f486b8905c50329722beb70fd43a71d/internal/util.go#L63-L67

This does not allow negative durations, but days, weeks, months or years. Thus, you might want to adjust both the pattern and title to the one in this method's documentation.

File: internal/server.go [hunk 4]
```diff
@@ -11,50 +11,142 @@ import (
...
+
+  This software stores the IP address for each upload. This information is
+  stored as long as the file is available. In addition, the IP address of the
+  user might be loged in case of an error. A normal download is logged without
+  user information.
+
+  <h2>## Abuse</h2>
+
+  If, for whatever reason, you would like to have a file removed prematurely,
+  please write an e-mail to <{{.EMail}}>. Please allow me a certain amount of
```
Comment (oxzi) [inline internal/server.go:155]: The `.EMail` field will result in an invalid HTML tag, e.g., `foo@example.org` results in `…please write an e-mail to <foo@example.org>.`. You might want to change this either to a `mailto` link or `<pre>` the e-mail address?

**Review by oxzi — CHANGES_REQUESTED** [review]
Thank you very much for this Pull Request. I have noted a few little things. Would you like to adjust them?

Once again, thank you very much.

**Comment by oxzi** [comment]
The review was The review referred the previous version. I'm going to have a look at the just updated version again.

File: internal/server.go [hunk 5]
```diff
@@ -11,50 +11,151 @@ import (
...
+  <form
+	action="{{.Proto}}://{{.Hostname}}/"
+	method="POST"
+	enctype="multipart/form-data"
+  >
+	<div id="grid">
+	  <label for="file">Your file:</label>
+	  <input type="file" name="file" />
+	  <label for="burn">Burn after reading:</label>
+	  <input
```
Comment (oxzi) [inline internal/server.go:126]: Wouldn't a checkbox here be more intuitive?

File: internal/server.go [hunk 6]
```diff
@@ -11,50 +11,151 @@ import (
...
+		type="text"
+		name="burn"
+		pattern="1?"
+		title="Use '1' to burn, leave empty to retain your file"
+	  />
+	  <label for="time">Optionally, set a custom expiry date:</label>
+	  <input
```
Comment (oxzi) [inline internal/server.go:133]: The server does not parse a time.Duration, only a variant.

https://github.com/oxzi/gosh/blob/8185d2179f486b8905c50329722beb70fd43a71d/internal/util.go#L63-L67

This does not allow negative durations, but days, weeks, months or years. Thus, you might want to adjust both the pattern and title to the one in this method's documentation.

**Review by oxzi — CHANGES_REQUESTED** [review]
I have added two old comments, which were marked as outdated due to the changes.

File: internal/server.go (hunk 2 above)
Comment (oxzi) [reply internal/server.go]: Addressed in the updated review.

File: internal/server.go (hunk 3 above)
Comment (oxzi) [reply internal/server.go]: Addressed in the updated review.

Comment (sgfk) [reply internal/server.go]: Q: So what would be valid time strings? Is "1mo 3d" valid, or only single dateinterval terms like "1mo" or "1d" on their own?

Comment (oxzi) [reply internal/server.go]: Any ordered sequence of duration pairs defined by a positive decimal number and a time unit specifier, e.g., `1h30m` or `23m42s`.

All possible time unit specifiers are defined in `internal/util.go`.
https://github.com/oxzi/gosh/blob/8185d2179f486b8905c50329722beb70fd43a71d/internal/util.go#L28-L38

Furthermore, some examples are available in the test. Please notice third `valid` field which indicates if one testcase represents a valid duration.
https://github.com/oxzi/gosh/blob/8185d2179f486b8905c50329722beb70fd43a71d/internal/util_test.go#L19-L29

Finally, the `getDurationPattern` function generates a regular expression for such a string.
https://github.com/oxzi/gosh/blob/8185d2179f486b8905c50329722beb70fd43a71d/internal/util.go#L45-L61

**Review by sgfk — COMMENTED** [review]
@oxzi Thank you so much for your review. I totally agree with your comments and suggestions.

**TODO**

- [ ] convert `getDurationPattern`-Result to string when passed to template

**Review by oxzi — APPROVED** [review]
LGTM, thanks!

**Comment by oxzi** [comment]
Thanks a lot for your work on adding this feature!

**Pull request merged**